<a href="https://colab.research.google.com/github/apexro45/WEEK-5/blob/main/WEEK_5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [25]:
!pip install db-sqlite3
import pandas as pd
import sqlite3 as sql

  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 172.0/172.0 kB 2.1 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for db-sqlite3: filename=db_sqlite3-0.0.1-py3-none-any.whl size=1770 sha256=3c38f154fc6fc85c58d47cfd84da936cf54c9a614309a86d702ce20fd0dac016
  Stored in directory: /root/.cache/pip/wheels/a6/b7/83/e941e0a0e04f417982e718ae7295d1e82b5f2863a1c51edd71
  Created wheel for db: filename=db-0.1.1-py3-none-any.whl size=3874 sha256=1d03572a29c7c294c76530d684a86005a58d335276f7b0f09b09ee75d1617375
  Stored in directory: /root/.cache/pip/wheels/7d/e4/df/bc55b93af204ab098d9effec76f6889ad12d7ad74e833c4910
  Created wheel for antiorm: filename=antiorm-1.2.1-py3-none-any.whl size=31665 sha256=b3a0371bff4a38e4077beef3d37f7d1114142e908d7cd058d21fdfe1ef39b753
  Stored in directory: /root/.cache/pip/wheels/30/9f/7e/b7c95b391cfa77a9e722d359e9c669cf6c8d798d748aec5091
Successfully built db-sq

In [26]:
from google.colab import files

uploaded = files.upload()

Saving sql-murder-mystery.db to sql-murder-mystery (1).db


In [58]:
db_file_name = 'sql-murder-mystery.db'
connection = sqlite3.connect(db_file_name)

cursor = connection.cursor()

cursor.execute("SELECT name FROM sqlite_master WHERE type='table';")
tables = cursor.fetchall()

for table in tables:
    print(table[0])
cursor.close()

connection.commit()



crime_scene_report
drivers_license
person
facebook_event_checkin
interview
get_fit_now_member
get_fit_now_check_in
income
solution


In [59]:
query_1 = '''
SELECT *
FROM crime_scene_report
WHERE city = "SQL City"
ORDER BY date;
'''
pd.set_option('display.max_colwidth', None)
pd.read_sql_query(query_1, connection)

date     type  \
0  20170712    theft   
1  20170820    arson   
2  20171110  robbery   
3  20180103  bribery   
4  20180115  assault   
5  20180115  assault   
6  20180115   murder   
7  20180215   murder   
8  20180215   murder   

                                                                                                                                                                                                                                                              description  \
0              A lone hunter stalks the night, firing arrows into the Darkness.\n            There is no hiding, no escape. In the distance, the beast\n            falters, tethered to the void. The killing blow comes without\n            hesitation, without mercy.   
1                                                                                                                                                                  Wield the Hammer of Sol with honor, Titan, it is a thing of\n            legend, both past and future.   
2  The Gjallarhorn shoulder-mounted rocket system was forged from\n            the armor of Guardians who fell at the Twilight Gap. Gifted\n            to the survivors of that terrible battle, the Gjallarhorn\n            is seen as a symbol of honor and survival.   
3                                                                                                                                                 Apparently, Cayde thought it necessary to expose this extremely\n            rare vegetable to a Hive summoning ritual.   
4                                                                                                                                                                                             Hamilton: Lee, do you yield? Burr: You shot him in the side! Yes he yields!   
5                                                                                                                                                                                                                                                        Report Not Found   
6                                                                               Security footage shows that there were 2 witnesses. The first witness lives at the last house on "Northwestern Dr". The second witness, named Annabel, lives somewhere on "Franklin Ave".   
7                                                                                                                                                                                                                                              REDACTED REDACTED REDACTED   
8                                                                                                                                                                                                                 Someone killed the guard! He took an arrow to the knee!   

       city  
0  SQL City  
1  SQL City  
2  SQL City  
3  SQL City  
4  SQL City  
5  SQL City  
6  SQL City  
7  SQL City  
8  SQL City

In [68]:
query_2 = '''
SELECT *
FROM Person
WHERE address_street_name ='Northwestern Dr'
ORDER BY address_number desc;
'''
pd.read_sql_query(query_2, connection)[:1]

id            name  license_id  address_number address_street_name  \
0  14887  Morty Schapiro      118009            4919     Northwestern Dr   

         ssn  
0  111564949

In [70]:
query_3 = '''
SELECT *
FROM Person
WHERE address_street_name ='Franklin Ave'
ORDER BY address_number desc;
'''
pd.read_sql_query(query_3, connection)[:10]

id                name  license_id  address_number address_street_name  \
0  24737          Gema Nantz      273410            3968        Franklin Ave   
1  84531      Edgar Mendieta      315790            3881        Franklin Ave   
2  36328        Domenic Laun      971080            3809        Franklin Ave   
3  34592   Cordell Lindamood      592762            3657        Franklin Ave   
4  83003      Wilmer Casella      672050            3564        Franklin Ave   
5  78408        Dante Eschen      218173            3447        Franklin Ave   
6  70050  Whitney Herkenratt      138462            3445        Franklin Ave   
7  36400        Noe Pancoast      895755            3443        Franklin Ave   
8  78170         Graig Summy      192325            3387        Franklin Ave   
9  78830       Candice Train      814779            3255        Franklin Ave   

         ssn  
0  180545802  
1  458847132  
2  825329637  
3  509890333  
4  569057540  
5  135374633  
6  552518972  
7  732212073  
8  277793306  
9  717570815

In [71]:
query_4 = '''
SELECT *
FROM interview
WHERE person_id = 14887 OR person_id = 16371;
'''
pd.read_sql_query(query_4, connection)

person_id  \
0      14887   
1      16371   

                                                                                                                                                                                                                        transcript  
0  I heard a gunshot and then saw a man run out. He had a "Get Fit Now Gym" bag. The membership number on the bag started with "48Z". Only gold members have those bags. The man got into a car with a plate that included "H42W".  
1                                                                                                            I saw the murder happen, and I recognized the killer from my gym when I was working out last week on January the 9th.

In [72]:
query_5 = '''
SELECT *
FROM get_fit_now_check_in
WHERE membership_id like "%48Z%" AND check_in_date = 20180109
order by check_in_date;
'''
pd.read_sql_query(query_5, connection)

membership_id  check_in_date  check_in_time  check_out_time
0         48Z7A       20180109           1600            1730
1         48Z55       20180109           1530            1700

In [73]:
query_6 = '''
SELECT *
FROM drivers_license
WHERE plate_number like "%H42W%";
'''
pd.read_sql_query(query_6, connection)

id  age  height eye_color hair_color  gender plate_number   car_make  \
0  183779   21      65      blue     blonde  female       H42W0X     Toyota   
1  423327   30      70     brown      brown    male       0H42W2  Chevrolet   
2  664760   21      71     black      black    male       4H42WR     Nissan   

  car_model  
0     Prius  
1  Spark LS  
2    Altima



Two males with a plate  (H42W)


In [78]:
query_7 = '''
SELECT *
FROM person
WHERE license_id = "423327" OR license_id = "664760";
'''
pd.read_sql_query(query_7, connection)

id            name  license_id  address_number    address_street_name  \
0  51739  Tushar Chandra      664760             312                 Phi St   
1  67318   Jeremy Bowers      423327             530  Washington Pl, Apt 3A   

         ssn  
0  137882671  
1  871539279

In [79]:
query_8 = '''
SELECT *
FROM get_fit_now_member
WHERE person_id = "51739" OR person_id = "67318";
'''
pd.read_sql_query(query_8, connection)

id  person_id           name  membership_start_date membership_status
0  48Z55      67318  Jeremy Bowers               20160101              gold

Google Colab, only allows the execution of one statement at a time

In [87]:

db_file_name = 'sql-murder-mystery.db'


connection = sqlite3.connect(db_file_name)


cursor = connection.cursor()


cursor.execute("INSERT INTO solution VALUES (1, 'Jeremy Bowers');")


connection.commit()


cursor.execute("SELECT value FROM solution;")


result = cursor.fetchone()


print("Result:", result[0] if result else "No result")




Result: Congrats, you found the murderer! But wait, there's more... If you think you're up for a challenge, try querying the interview transcript of the murderer to find the real villain behind this crime. If you feel especially confident in your SQL skills, try to complete this final step with no more than 2 queries. Use this same INSERT statement with your new suspect to check your answer.


In [88]:
query_9 = '''
SELECT *
FROM interview
WHERE person_id = 67318;
'''
pd.read_sql_query(query_9, connection)

person_id  \
0      67318   

                                                                                                                                                                                                                                           transcript  
0  I was hired by a woman with a lot of money. I don't know her name but I know she's around 5'5" (65") or 5'7" (67"). She has red hair and she drives a Tesla Model S. I know that she attended the SQL Symphony Concert 3 times in December 2017.\n

In [89]:
query_10 = '''
SELECT *
FROM drivers_license
WHERE car_make = "Tesla" AND car_model = "Model S" AND
gender = "female" AND hair_color = "red";
'''
pd.read_sql_query(query_10, connection)

id  age  height eye_color hair_color  gender plate_number car_make  \
0  202298   68      66     green        red  female       500123    Tesla   
1  291182   65      66      blue        red  female       08CM64    Tesla   
2  918773   48      65     black        red  female       917UU3    Tesla   

  car_model  
0   Model S  
1   Model S  
2   Model S

so there are 3 ppl with model s car and red hair

In [90]:
query_11 = '''
SELECT *
FROM person
WHERE license_id = "202298" OR license_id = "291182" OR license_id = "918773";
'''
pd.read_sql_query(query_11, connection)

id              name  license_id  address_number address_street_name  \
0  78881          Red Korb      918773             107         Camerata Dr   
1  90700     Regina George      291182             332           Maple Ave   
2  99716  Miranda Priestly      202298            1883          Golden Ave   

         ssn  
0  961388910  
1  337169072  
2  987756388

In [92]:
query_12 = '''
SELECT person_id, count(*), event_name
FROM facebook_event_checkin
GROUP BY person_id
having count(*) = 3 AND event_name = "SQL Symphony Concert" AND date like "%201712%";
'''
pd.read_sql_query(query_12, connection)

person_id  count(*)            event_name
0      24556         3  SQL Symphony Concert
1      99716         3  SQL Symphony Concert

99716 	matches - Miranda Priestly is the brains behind the murder!

